components/brave_wallet/browser/nft_metadata_fetcher.cc

```diff
@@ -0,0 +1,376 @@
+    std::move(callback).Run(
+        "", static_cast<int>(mojom::JsonRpcError::kParsingError),
+        l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+    return;
+  }
+
+  std::move(callback).Run(api_request_result.body(), 0, "");  // 0 is kSuccess
+}
+
+void NftMetadataFetcher::CompleteGetEthTokenMetadata(
+    GetEthTokenMetadataCallback callback,
+    const std::string& response,
+    int error,
+    const std::string& error_message) {
+  std::move(callback).Run(response, mojom::ProviderError(error), error_message);
+}
+
+void NftMetadataFetcher::GetSolTokenMetadata(
+    const std::string& nft_account_address,
+    GetSolTokenMetadataCallback callback) {
+  // Derive metadata PDA for the NFT accounts
+  absl::optional<std::string> associated_metadata_account =
+      SolanaKeyring::GetAssociatedMetadataAccount(nft_account_address);
+  if (!associated_metadata_account) {
+    std::move(callback).Run(
+        "", mojom::SolanaProviderError::kInternalError,
+        l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+    return;
+  }
+
+  auto internal_callback =
+      base::BindOnce(&NftMetadataFetcher::OnGetSolanaAccountInfoTokenMetadata,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback));
+  json_rpc_service_->GetSolanaAccountInfo(*associated_metadata_account,
+                                          std::move(internal_callback));
+}
+
+void NftMetadataFetcher::OnGetSolanaAccountInfoTokenMetadata(
+    GetSolTokenMetadataCallback callback,
+    absl::optional<SolanaAccountInfo> account_info,
+    mojom::SolanaProviderError error,
+    const std::string& error_message) {
+  if (error != mojom::SolanaProviderError::kSuccess || !account_info) {
+    std::move(callback).Run("", std::move(error), error_message);
+    return;
+  }
+
+  absl::optional<std::vector<uint8_t>> metadata =
+      base::Base64Decode((*account_info).data);
+  if (!metadata) {
+    std::move(callback).Run("", mojom::SolanaProviderError::kParsingError,
+                            l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+    return;
+  }
+
+  absl::optional<GURL> url = DecodeMetadataUri(*metadata);
+  if (!url || !url.value().is_valid()) {
+    std::move(callback).Run("", mojom::SolanaProviderError::kParsingError,
+                            l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+    return;
+  }
+
+  FetchMetadata(*url, base::BindOnce(
+                          &NftMetadataFetcher::CompleteGetSolTokenMetadata,
+                          weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void NftMetadataFetcher::CompleteGetSolTokenMetadata(
+    GetSolTokenMetadataCallback callback,
+    const std::string& response,
+    int error,
+    const std::string& error_message) {
+  std::move(callback).Run(response, mojom::SolanaProviderError(error),
+                          error_message);
+}
+
+// static
+absl::optional<uint32_t> NftMetadataFetcher::DecodeUint32(
```

Comment:
If this is just a copy of [this function](https://github.com/brave/brave-core/blob/master/components/brave_wallet/browser/solana_instruction_data_decoder.cc#L629) do we need to put it here as well?

Reply:
Yes, this is not ideal.  I was resolving this feedback https://github.com/brave/brave-core/pull/15914#issuecomment-1334250779, to separate the code.

I think the issue is that it's not clear which code from  components/brave_wallet/browser/solana_instruction_data_decoder.cc is reusable for borsh decoding and which isn't.  This function is, but the `DecodeString` function was not, for example.

I think the ideally we determine which code is in common, and move that to a shared util, but I am not sure if that's in scope for this change cc @yrliou

Reply:
I'm good leaving this out of scope and handling this in a refactor. Particularly just because A, I see this is a P2 so we need to get it shipped, B I've looked over it and am pretty certain we're not introducing any memory corruption vulnerabilities, and C I think it would be good for us to backtrack and do a review of all of these functions now that I'm seeing how this works to make sure we didn't accidentally create a memory corruption issue here.